src/main/java/tc/oc/pgm/commands/MapCommands.java

```diff
@@ -62,12 +66,19 @@ public static void maplist(
       throws CommandException {
     Stream<MapInfo> search = Sets.newHashSet(library.getMaps()).stream();
     if (tags != null) {
-      final Set<String> tagSet =
+      final Map<Boolean, List<String>> tagSet =
           Stream.of(tags.split(","))
               .map(String::toLowerCase)
               .map(String::trim)
-              .collect(Collectors.toSet());
-      search = search.filter(map -> matchesTags(map, tagSet));
+              .collect(Collectors.partitioningBy(s -> s.startsWith("!")));
+      search =
+          search.filter(
+              map ->
+                  matchesTags(
+                      map,
+                      tagSet.getOrDefault(false, new LinkedList<>()),
+                      Collections2.transform(
+                          tagSet.getOrDefault(true, new LinkedList<>()), s -> s.substring(1))));
```

Comment:
should probably just use:

```
        final Map<Boolean, Set<String>> tagSet = Stream.of(tags.split(","))
                .map(String::toLowerCase)
                .map(String::trim)
                .collect(Collectors.partitioningBy(s -> s.startsWith("!"),
                        Collectors.mapping(
                                (String s) ->  s.startsWith("!") ? s.substring(1) : s,
                                Collectors.toSet())));
```
that does both partitioning and removing the first char, that way you avoid the later transformation on the filter.

Also, i don't think linked lists are a good fit, should probably just use hash sets.